### DISCUSSION TITLE
Debugging with vscode integrated debugger

### DISCUSSION THREAD

**Discussion by fantauzzi** [smartcontractkit/full-blockchain-solidity-course-py#1617] [General]
It is possible to configure the **vscode debugger** to run and debug Brownie scripts, including unit testing. I find the vscode debugger more productive and user-friendly than `pdb`. 

Here how the vscode debugger can be configured to work with Brownie scripts and unit-tests.

- In vscode, open the directory with the project of interest
- Click on the `Run and Debug` button on the left bar (or the menu `View > Run`)
- Pull down the `RUN AND DEBUG` menu (top-left) and choose `Add Configuration...`
- A `launch.json` should open, also a menu, press `Esc` to close the menu without choosing anything from it
- Copy and paste the script below into the `launch.json`. If you already have in there previous debugging configurations that you want to keep, copy only the `configurations` section from the script below into the corresponding section of your `launch.json`
- Everywhere in the script where there is a `program` entry, you must set its value to the complete (absolute) path to your `brownie` executable, replacing the path I am using (`/home/fanta/.local/virtualenv/python3.10/bin/brownie`)

Note: to find where your `brownie` executable is, you can run in the shell:
`which brownie`

To debug:
- set any breakpoint you want in your source code (click to the left of the line number)
- click on the `Run and Debug` button on the left (or the menu `View > Run`)
- open the `RUN AND DEBUG` pull-down menu and choose a configuration from it
- start the script/test with the debugger by clicking on the `▷` button next to `RUN AND DEBUG`.

Note that is is possible to issue commands in Python in the debug console, pretty much like from the Brownie console, but with the benefit of being in the execution context of the scripts, with access to their functions, classes, variables and imported packages. 

It is quite easy to customize/improve the `launch.json`; taking a gander at it, you can see how to add entries to the `RUN AND DEBUG` menu, how to add other networks to choose from, options to run Brownie, etc.

Here the `launch.json` script

```
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    /*{
              "name": "Brownie: Current File",
              "type": "python",
              "request": "launch",
              "program": "${file}",
              "console": "integratedTerminal",
              "justMyCode": true
          }*/
    {
      "name": "Brownie: run deploy.js",
      "type": "python",
      "request": "launch",
      "program": "/home/fanta/.local/virtualenv/python3.10/bin/brownie",
      "console": "integratedTerminal",
      "args": ["run", "scripts/deploy.js"]
    },
    {
      "name": "Brownie: run deploy.js in a network",
      "type": "python",
      "request": "launch",
      "program": "/home/fanta/.local/virtualenv/python3.10/bin/brownie",
      "console": "integratedTerminal",
      "args": ["run", "scripts/deploy.js", "--network", "${input:network}"]
    },

    {
      "name": "Brownie: run current file",
      "type": "python",
      "request": "launch",
      "program": "/home/fanta/.local/virtualenv/python3.10/bin/brownie",
      "console": "integratedTerminal",
      "args": ["run", "${file}"]
    },
    {
      "name": "Brownie: run current file in a network",
      "type": "python",
      "request": "launch",
      "program": "/home/fanta/.local/virtualenv/python3.10/bin/brownie",
      "console": "integratedTerminal",
      "args": ["run", "${file}", "--network", "${input:network}"]
    },
    {
      "name": "Brownie: test",
      "type": "python",
      "request": "launch",
      "program": "/home/fanta/.local/virtualenv/python3.10/bin/brownie",
      "console": "integratedTerminal",
      "args": ["test", "-v"]
    },
    {
      "name": "Brownie: test in network",
      "type": "python",
      "request": "launch",
      "program": "/home/fanta/.local/virtualenv/python3.10/bin/brownie",
      "console": "integratedTerminal",
      "args": ["test", "-v", "--network", "${input:network}"]
    },
    {
      "name": "Brownie: test -k <test_function>",
      "type": "python",
      "request": "launch",
      "program": "/home/fanta/.local/virtualenv/python3.10/bin/brownie",
      "console": "integratedTerminal",
      "args": ["test", "-v", "-k", "${input:function}"]
    }
  ],
  "inputs": [
    {
      "id": "network",
      "type": "pickString",
      "description": "network",
      "options": [
        "development",
        "ganache-local",
        "mainnet-fork-dev",
        "kovan",
        "rinkeby"
      ],
      "default": "development"
    },
    {
      "id": "function",
      "type": "promptString",
      "description": "Test function"
    }
  ]
}
```

**Comment by cromewar** [comment]
Awesome work @fantauzzi !!
